[Python install] ld: library not found for -latomic

### What version of gRPC and what language are you using?

gRPC 1.27.2, Python

### What operating system (Linux, Windows,...) and version?

macOS 10.15.3

### What runtime / compiler are you using (e.g. python version or version of gcc)

Python 3.7.6, Clang 11.0.3

### What did you do?

Building from source, using `python setup.py build && python setup.py install` with the following environment variables:
```bash
GRPC_PYTHON_BUILD_WITH_CYTHON=True
GRPC_PYTHON_BUILD_SYSTEM_OPENSSL=True
GRPC_PYTHON_BUILD_SYSTEM_ZLIB=True
GRPC_PYTHON_BUILD_SYSTEM_CARES=True
```

### What did you expect to see?

A successful build.

### What did you see instead?

```
ld: library not found for -latomic
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

### Anything else we should know about your project / environment?

I'm actually using the [Spack](https://spack.io) package manager to build gRPC from source. Here are the relevant logs:

* [spack-build-out.txt](https://github.com/grpc/grpc/files/4389896/spack-build-out.txt)
* [spack-build-env.txt](https://github.com/grpc/grpc/files/4389897/spack-build-env.txt)

I believe this issue is related to #20414, so @veblush may be able to help. Either `check_linker_need_libatomic` behaves incorrectly, or maybe I just need to build libatomic myself.

Looking further at the code, I think I just need to install libatomic on my system:
```console
$ cat test.cxx
#include <atomic>
int main() { return std::atomic<int64_t>{}; }
$ clang -x c++ -std=c++11 test.cxx 
Undefined symbols for architecture x86_64:
  "std::terminate()", referenced from:
      ___clang_call_terminate in test-1858b3.o
  "___cxa_begin_catch", referenced from:
      ___clang_call_terminate in test-1858b3.o
  "___gxx_personality_v0", referenced from:
      std::__1::__atomic_base<long long, false>::load(std::__1::memory_order) const in test-1858b3.o
      Dwarf Exception Unwind Info (__eh_frame) in test-1858b3.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
$ clang -x c++ -std=c++11 -latomic test.cxx 
ld: library not found for -latomic
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I'm seeing the same with a simple `pip install grcpio` on Python 3.9.0b3 (but it doesn't look like it's related to the Python version).

I tried to compile the test program used by `check_linker_need_libatomic` manually and this is what I get:

```
$ cat test.cpp | cc -x c++ -std=c++11 -                                                                                    (master) (tienda-3.8.2)
Undefined symbols for architecture x86_64:
  "std::terminate()", referenced from:
      ___clang_call_terminate in --dd656e.o
  "___cxa_begin_catch", referenced from:
      ___clang_call_terminate in --dd656e.o
  "___gxx_personality_v0", referenced from:
      std::__1::__atomic_base<long long, false>::load(std::__1::memory_order) const in --dd656e.o
      Dwarf Exception Unwind Info (__eh_frame) in --dd656e.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Based on random searching I found https://github.com/pybind/pybind11/issues/382 and the solution suggested there seems to work

Adding `-lc++` to the test invocation in `check_linker_need_libatomic` seems to solve the problem on Catalina 10.15.5.

original error:
```
$ echo -e "#include <atomic>\nint main() { return std::atomic<int64_t>{}; }" | cc -x c++ -std=c++11 -
Undefined symbols for architecture x86_64:
  "std::terminate()", referenced from:
      ___clang_call_terminate in --905908.o
  "___cxa_begin_catch", referenced from:
      ___clang_call_terminate in --905908.o
  "___gxx_personality_v0", referenced from:
      std::__1::__atomic_base<long long, false>::load(std::__1::memory_order) const in --905908.o
      Dwarf Exception Unwind Info (__eh_frame) in --905908.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

with additional flag:
```
$ echo -e "#include <atomic>\nint main() { return std::atomic<int64_t>{}; }" | cc -x c++ -std=c++11 -lc++ -
$
```

--- a/setup.py
+++ b/setup.py
@@ -166,22 +166,22 @@ def check_linker_need_libatomic():
     """Test if linker on system needs libatomic."""
     code_test = (b'#include <atomic>\n' +
                  b'int main() { return std::atomic<int64_t>{}; }')
-    cc_test = subprocess.Popen(['cc', '-x', 'c++', '-std=c++11', '-'],
-                               stdin=PIPE,
-                               stdout=PIPE,
-                               stderr=PIPE)
-    cc_test.communicate(input=code_test)
-    if cc_test.returncode == 0:
+    cpp_test = subprocess.Popen(['c++', '-x', 'c++', '-std=c++11', '-'],
+                                stdin=PIPE,
+                                stdout=PIPE,
+                                stderr=PIPE)
+    cpp_test.communicate(input=code_test)
+    if cpp_test.returncode == 0:
         return False
     # Double-check to see if -latomic actually can solve the problem.
     # https://github.com/grpc/grpc/issues/22491
-    cc_test = subprocess.Popen(
-        ['cc', '-x', 'c++', '-std=c++11', '-latomic', '-'],
+    cpp_test = subprocess.Popen(
+        ['c++', '-x', 'c++', '-std=c++11', '-latomic', '-'],
         stdin=PIPE,
         stdout=PIPE,
         stderr=PIPE)
-    cc_test.communicate(input=code_test)
-    return cc_test.returncode == 0
+    cpp_test.communicate(input=code_test)
+    return cpp_test.returncode == 0
 
 
 # There are some situations (like on Windows) where CC, CFLAGS, and LDFLAGS are

--- a/setup.py
+++ b/setup.py
@@ -171,7 +171,17 @@ def check_linker_need_libatomic():
                                stdout=PIPE,
                                stderr=PIPE)
     cc_test.communicate(input=code_test)
-    return cc_test.returncode != 0
+    if cc_test.returncode == 0:
+        return False
+    # Double-check to see if -latomic actually can solve the problem.
+    # https://github.com/grpc/grpc/issues/22491
+    cc_test = subprocess.Popen(
+        ['cc', '-x', 'c++', '-std=c++11', '-latomic', '-'],
+        stdin=PIPE,
+        stdout=PIPE,
+        stderr=PIPE)
+    cc_test.communicate(input=code_test)
+    return cc_test.returncode == 0
 
 
 # There are some situations (like on Windows) where CC, CFLAGS, and LDFLAGS are

